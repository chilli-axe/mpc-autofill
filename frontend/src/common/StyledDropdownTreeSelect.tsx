import DropdownTreeSelect from "react-dropdown-tree-select";
import styled from "styled-components";

export const StyledDropdownTreeSelect = styled(DropdownTreeSelect)`
  .tag {
    color: black;
    background-color: #dddddd;
  }
  .tag-remove {
    color: #666666;
  }

  .dropdown-trigger {
    border-radius: 0.25rem;
    background-color: white;
  }
  .dropdown-content {
    border-radius: 0.25rem;
  }

  .search {
    background-color: white;
    color: black;
  }
  .search::placeholder {
    color: black;
  }
  .placeholder {
    color: gray;
  }

  .toggle {
    font: normal normal normal 12px/1 bootstrap-icons;
    top: 2px;
    left: 2px;
  }

  .toggle.collapsed::after {
    content: "\F4FA";
  }

  .toggle.expanded::after {
    content: "\F2E6";
  }

  color: black;

  .root {
    padding: 0;
    margin: 0;
  }
`;
