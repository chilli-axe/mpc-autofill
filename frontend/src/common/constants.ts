import { CardType as CardTypeSchema } from "@/common/schema_types";
import { Cardstock, CardType, Faces, SortBy } from "@/common/types";
export const ProjectName = "MPC Autofill";
export const MakePlayingCards = "MakePlayingCards.com";
export const MakePlayingCardsURL = "https://www.makeplayingcards.com";

export const Card: CardType = CardTypeSchema.Card;
export const Cardback: CardType = CardTypeSchema.Cardback;
export const Token: CardType = CardTypeSchema.Token;

export const SelectedImageSeparator = "@";
export const CardTypeSeparator = ":";
export const FaceSeparator = "//";

export const CardTypePrefixes: { [prefix: string]: CardType } = {
  "": Card,
  b: Cardback,
  t: Token,
};

export const ReversedCardTypePrefixes = Object.fromEntries(
  Object.keys(CardTypePrefixes).map((prefix: string) => [
    CardTypePrefixes[prefix],
    prefix.length > 0 ? prefix + CardTypeSeparator : prefix,
  ])
);

export const Front: Faces = "front";
export const Back: Faces = "back";

export const ToggleButtonHeight = 38; // pixels
export const NavbarHeight = 50; // pixels - aligns with the natural height of the navbar
export const RibbonHeight = 54; // pixels
export const NavbarLogoHeight = 40; // pixels
export const ContentMaxWidth = 1200; // pixels - aligns with bootstrap's large breakpoint

export const MinimumDPI = 0;
export const MaximumDPI = 1500;
export const DPIStep = 50;

export const MaximumSize = 30; // megabytes
export const SizeStep = 1;

export const CSRFKey = "csrftoken";
export const SearchSettingsKey = "searchSettings";
export const GoogleAnalyticsConsentKey = "googleAnalyticsConsent";
export const BackendURLKey = "backendURL";

export const Brackets: Array<number> = [
  18, 36, 55, 72, 90, 108, 126, 144, 162, 180, 198, 216, 234, 396, 504, 612,
];

export const ProjectMaxSize: number = Brackets[Brackets.length - 1];

export enum QueryTags {
  BackendSpecific = "backendSpecific",
  SearchResults = "searchResults",
  SampleCards = "sampleCards",
}

export const S27: Cardstock = "(S27) Smooth";
export const S30: Cardstock = "(S30) Standard Smooth";
export const S33: Cardstock = "(S33) Superior Smooth";
export const M31: Cardstock = "(M31) Linen";
export const P10: Cardstock = "(P10) Plastic";
export const Cardstocks: Array<Cardstock> = [S27, S30, S33, M31, P10];

export const CardstockFoilCompatibility: { [cardstock in Cardstock]: boolean } =
  {
    [S27]: true,
    [S30]: true,
    [S33]: true,
    [M31]: true,
    [P10]: false,
  };

export const GoogleDriveImageAPIURL =
  "https://script.google.com/macros/s/AKfycbw8laScKBfxda2Wb0g63gkYDBdy8NWNxINoC4xDOwnCQ3JMFdruam1MdmNmN4wI5k4/exec";

export const SearchResultsEndpointPageSize = 300;
export const CardEndpointPageSize = 1000;

export enum CSVHeaders {
  quantity = "Quantity",
  frontQuery = "Front",
  frontSelectedImage = "Front ID",
  backQuery = "Back",
  backSelectedImage = "Back ID",
}

export const ExploreDebounceMS = 700;
export const ExplorePageSize = 60;

export const SortByOptions: { [option in SortBy]: string } = {
  dateCreatedDescending: "Date Created (Newest-Oldest)",
  dateCreatedAscending: "Date Created (Oldest-Newest)",
  dateModifiedDescending: "Date Modified (Newest-Oldest)",
  dateModifiedAscending: "Date Modified (Oldest-Newest)",
  nameAscending: "Name (A-Z)",
  nameDescending: "Name (Z-A)",
};
