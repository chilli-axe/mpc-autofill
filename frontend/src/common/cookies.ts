/**
 * Data API for interacting with anything stored in cookies or local storage.
 */

// import Ajv2020 from "ajv/dist/2020";
import Cookies from "js-cookie";

import {
  BackendURLKey,
  CSRFKey,
  GoogleAnalyticsConsentKey,
  SearchSettingsKey,
} from "@/common/constants";
import { Convert } from "@/common/schema_types";
import { SearchSettings, SourceDocuments, SourceRow } from "@/common/types";
import { getSourceRowsFromSourceSettings } from "@/common/utils";
import { getDefaultSearchSettings } from "@/store/slices/searchSettingsSlice";

// import * as SearchSettingsSchema from "../../../schemas/schemas/SearchSettings.json";

// const ajv = new Ajv2020();

//# region CSRF
// TODO: unsure if we still need this.

export function getCSRFHeader(): HeadersInit | undefined {
  const csrfToken = Cookies.get(CSRFKey);
  if (csrfToken != null) {
    return { "X-CSRFToken": csrfToken };
  }
  return undefined;
}

//# endregion

//# region search settings

/**
 * Get search settings from localStorage data. If valid data is retrieved,
 * ensure that all `sourceDocuments` are included in the returned settings,
 * with any new sources that weren't previously included added to the end and enabled.
 */
export function getLocalStorageSearchSettings(
  sourceDocuments: SourceDocuments
): SearchSettings {
  const serialisedRawSettings = localStorage.getItem(SearchSettingsKey) ?? "{}";
  try {
    const searchSettings = Convert.toSearchSettings(serialisedRawSettings);
    // great, the user has valid search settings stored in their browser local storage.
    // reconcile against sourceDocuments
    const sourceInDatabaseSet: Set<number> = new Set(
      Object.values(sourceDocuments).map((sourceDocument) =>
        parseInt(sourceDocument.pk)
      )
    );
    // types have to be narrowed here because quicktype doesn't support our SourceRow data structure :(
    const sources: Array<SourceRow> = getSourceRowsFromSourceSettings(
      searchSettings.sourceSettings
    );
    const sourceInLocalStorageSet: Set<number> = new Set(
      sources.map((row) => row[0])
    );
    // one fat line of reconciliation, good luck reading this future nick! i wrote this at 12:26am.
    searchSettings.sourceSettings.sources = sources
      .filter((row: SourceRow) => sourceInDatabaseSet.has(row[0]))
      .concat(
        Array.from(sourceInDatabaseSet)
          .filter((pk: number) => !sourceInLocalStorageSet.has(pk))
          .map((pk: number) => [pk, true])
      );
    return searchSettings;
  } catch (e) {
    // quicktype will throw an error if the user's stored settings do not match the schema
    // e.g. upon first page load
    // just return the default settings
    return getDefaultSearchSettings(sourceDocuments);
  }
}

export function setLocalStorageSearchSettings(settings: SearchSettings): void {
  localStorage.setItem(SearchSettingsKey, JSON.stringify(settings));
}

//# endregion

//# region google analytics consent

export function getGoogleAnalyticsConsent(): boolean | undefined {
  const rawConsent = Cookies.get(GoogleAnalyticsConsentKey);
  return rawConsent != undefined ? JSON.parse(rawConsent) === true : undefined;
}

export function setGoogleAnalyticsConsent(consent: boolean): void {
  Cookies.set(GoogleAnalyticsConsentKey, JSON.stringify(consent), {
    expires: 365,
    sameSite: "strict",
  });
}

//# endregion

//# region backend

export function getLocalStorageBackendURL() {
  return localStorage.getItem(BackendURLKey);
}

export function setLocalStorageBackendURL(url: string) {
  localStorage.setItem(BackendURLKey, url);
}

export function clearLocalStorageBackendURL(): void {
  localStorage.removeItem(BackendURLKey);
}

//# endregion
