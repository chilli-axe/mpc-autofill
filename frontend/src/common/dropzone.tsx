/**
 * A small wrapper around the `react-dropzone` Dropzone for consistent presentation.
 */

import React, { useCallback } from "react";
import { DropzoneRootProps, useDropzone } from "react-dropzone";
import styled from "styled-components";

import { useAppDispatch } from "@/common/types";
import { setNotification } from "@/store/slices/toastsSlice";

const getColor = (props: DropzoneRootProps) => {
  if (props.isDragAccept) {
    return "#00e676";
  }
  if (props.isDragReject) {
    return "#ff1744";
  }
  if (props.isFocused) {
    return "#2196f3";
  }
  return "#eeeeee";
};

const Container = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
  border-width: 2px;
  border-radius: 2px;
  border-color: ${(props: DropzoneRootProps) => getColor(props)};
  border-style: dashed;
  color: #bdbdbd;
  outline: none;
  transition: border 0.24s ease-in-out;
  cursor: pointer;
`;

interface StyledDropzoneProps {
  mimeTypes: { [mimeType: string]: Array<string> };
  fileUploadCallback: {
    (fileContent: string | ArrayBuffer | null): void;
  };
  label: string;
  disabled: boolean;
}

export function TextFileDropzone({
  mimeTypes,
  fileUploadCallback,
  label,
  disabled,
}: StyledDropzoneProps) {
  const dispatch = useAppDispatch();

  // let the user know when something goes wrong with reading their file
  const onAbort = useCallback(
    () =>
      dispatch(
        setNotification([
          "dropzone-onabort",
          {
            name: "File upload error",
            message: "File reading was stopped.",
            level: "error",
          },
        ])
      ),
    [dispatch]
  );
  const onError = useCallback(
    () =>
      dispatch(
        setNotification([
          "dropzone-onerror",
          {
            name: "File upload error",
            message: "An error occurred while reading the file.",
            level: "error",
          },
        ])
      ),
    [dispatch]
  );
  const onDrop = useCallback(
    (acceptedFiles: Array<File>) => {
      acceptedFiles.forEach((file: File) => {
        const reader = new FileReader();
        reader.onabort = onAbort;
        reader.onerror = onError;
        reader.onload = () => {
          fileUploadCallback(reader.result);
        };
        reader.readAsText(file);
      });
    },
    [fileUploadCallback, onAbort, onError]
  );

  const { getRootProps, getInputProps, isFocused, isDragAccept, isDragReject } =
    useDropzone({ onDrop, accept: mimeTypes, maxFiles: 1, disabled });

  return (
    <div className="container">
      <Container
        {...getRootProps({ isFocused, isDragAccept, isDragReject })}
        aria-label={label ?? "dropzone"}
      >
        <input {...getInputProps()} />
        Drag and drop a file here, or click to select a file.
      </Container>
    </div>
  );
}
