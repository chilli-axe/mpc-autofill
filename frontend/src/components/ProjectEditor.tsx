/**
 * This component represents the complete MPC Autofill project editor, ready to
 * drop into a page (as the only component). Must be wrapped with a Redux provider.
 */

import React, { useEffect } from "react";
import Col from "react-bootstrap/Col";
import Row from "react-bootstrap/Row";
import styled from "styled-components";

import { RibbonHeight } from "@/common/constants";
import { useAppSelector } from "@/common/types";
import { NoBackendDefault } from "@/components/NoBackendDefault";
import { OverflowCol } from "@/components/OverflowCol";
import { Ribbon } from "@/components/Ribbon";
import { SelectedImagesRibbon } from "@/features/bulkManagement/SelectedImagesRibbon";
import { CardGrid } from "@/features/card/CardGrid";
import { CommonCardback } from "@/features/card/CommonCardback";
import { Export } from "@/features/export/Export";
import { FinishedMyProject } from "@/features/export/FinishedMyProjectModal";
import { FinishSettings } from "@/features/finishSettings/FinishSettings";
import { Import } from "@/features/import/Import";
import { SearchSettings } from "@/features/searchSettings/SearchSettings";
import { Status } from "@/features/status/Status";
import { useBackendConfigured } from "@/store/slices/backendSlice";
import {
  selectIsProjectEmpty,
  selectProjectCardback,
} from "@/store/slices/projectSlice";

function ProjectEditor() {
  // TODO: should we periodically ping the backend to make sure it's still alive?
  //# region queries and hooks

  const backendConfigured = useBackendConfigured();
  const cardback = useAppSelector(selectProjectCardback);
  const isProjectEmpty = useAppSelector(selectIsProjectEmpty);

  //# endregion

  //# region effects

  /**
   * Ask the user for confirmation before they close the page if their project has any cards in it.
   */
  useEffect(() => {
    const handler = (event: BeforeUnloadEvent) => {
      if (!isProjectEmpty) {
        event.preventDefault();
        return false;
      }
    };
    window.addEventListener("beforeunload", handler);
    return () => {
      window.removeEventListener("beforeunload", handler);
    };
  }, [isProjectEmpty]);

  //# endregion

  return (
    <>
      {backendConfigured ? (
        <Row className="g-0">
          <Ribbon className="g-0">
            <SelectedImagesRibbon />
          </Ribbon>
          <OverflowCol
            lg={8}
            md={8}
            sm={6}
            xs={6}
            data-testid="left-panel"
            heightDelta={RibbonHeight}
          >
            <CardGrid />
          </OverflowCol>
          <OverflowCol
            data-testid="right-panel"
            lg={4}
            md={4}
            sm={6}
            xs={6}
            style={{ zIndex: 1 }}
            className="px-2"
            heightDelta={RibbonHeight}
          >
            <Status />
            <Row className="g-0 pt-2">
              <FinishSettings />
            </Row>
            <Row className="g-0 pt-2">
              <SearchSettings />
            </Row>
            <Row className="g-0 pt-2">
              <Col lg={7} md={12} sm={12} xs={12}>
                <Import />
              </Col>
              <Col lg={5} md={12} sm={12} xs={12}>
                <Export />
              </Col>
              <FinishedMyProject />
            </Row>
            <Col className="g-0 pt-2" lg={{ span: 8, offset: 2 }} md={12}>
              <CommonCardback selectedImage={cardback} />
            </Col>
          </OverflowCol>
        </Row>
      ) : (
        <NoBackendDefault />
      )}
    </>
  );
}

export default ProjectEditor;
