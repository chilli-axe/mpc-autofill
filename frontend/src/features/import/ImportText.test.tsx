import { within } from "@testing-library/dom";
import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Back, Front, SelectedImageSeparator } from "@/common/constants";
import {
  cardDocument1,
  cardDocument2,
  cardDocument3,
  cardDocument4,
  cardDocument5,
  cardDocument6,
} from "@/common/test-constants";
import {
  expectCardbackSlotState,
  expectCardGridSlotState,
  expectCardSlotToExist,
  expectCardSlotToNotExist,
  importText,
  openImportTextModal,
  renderWithProviders,
} from "@/common/test-utils";
import ProjectEditor from "@/components/ProjectEditor";
import {
  cardbacksTwoOtherResults,
  cardDocumentsFourResults,
  cardDocumentsSixResults,
  cardDocumentsThreeResults,
  defaultHandlers,
  dfcPairsMatchingCards1And4,
  sampleCards,
  searchResultsForDFCMatchedCards1And4,
  searchResultsOneResult,
  searchResultsSixResults,
  sourceDocumentsOneResult,
  sourceDocumentsThreeResults,
} from "@/mocks/handlers";
import { server } from "@/mocks/server";

//# region mocks
// this ensures that the text import placeholder text is deterministic between test runs

beforeEach(() => {
  jest.spyOn(global.Math, "random").mockReturnValue(1);
});

afterEach(() => {
  jest.spyOn(global.Math, "random").mockRestore();
});

//# endregion

const preloadedState = {
  project: {
    members: [],
    cardback: cardDocument2.identifier,
    mostRecentlySelectedSlot: null,
  },
};

//# region snapshot tests

test("the html structure of text importer", async () => {
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await openImportTextModal();

  expect(screen.getByTestId("import-text")).toMatchSnapshot();
});

//# endregion

test("importing one card by text into an empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import a card
  await importText("my search query");

  // a card slot should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("importing multiple instances of one card by text into an empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import two instances of a card
  await importText("2x my search query");

  // two card slots should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(2, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("importing multiple instances of one card without an x by text into an empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  // import two instances of a card without an x
  await importText("2 my search query");

  // two card slots should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(2, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2);
});

test("importing multiple instances of one card with a capital X by text into an empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import two instances of a card with a capital X
  await importText("2X my search query");

  // two card slots should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(2, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("importing multiple instances of one card by text into a non-empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, {
    preloadedState: {
      ...preloadedState,
      project: {
        members: [],
        cardback: cardDocument2.identifier,
        mostRecentlySelectedSlot: null,
      },
    },
  });

  // this used to preload the redux state, but with the shift to listeners,
  // we have to add the first card manually like this.
  await importText(
    `1x my search query${SelectedImageSeparator}${cardDocument1.identifier}`
  );
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import two instances of a card
  await importText("2x my search query");

  // two card slots should have been created
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(2, Back, cardDocument2.name, 1, 2);
  await expectCardSlotToExist(3);
  await expectCardGridSlotState(3, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(3, Back, cardDocument2.name, 1, 2);
});

test("importing one card of each type into an empty project", async () => {
  server.use(
    cardDocumentsSixResults,
    cardbacksTwoOtherResults,
    sourceDocumentsThreeResults,
    searchResultsSixResults,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import one card of each type
  await importText("query 1\nt:query 6\nb:query 5");

  // three card slots should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument6.name, 1, 1);
  await expectCardGridSlotState(2, Back, cardDocument2.name, 1, 2);
  await expectCardSlotToExist(3);
  await expectCardGridSlotState(3, Front, cardDocument5.name, 1, 1);
  await expectCardGridSlotState(3, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("importing one DFC-paired card by text into an empty project", async () => {
  server.use(
    cardDocumentsFourResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsForDFCMatchedCards1And4,
    dfcPairsMatchingCards1And4,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import one instance of a double faced card
  await importText("my search query");

  // we should now have both sides of that DFC pair in slot 1
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument4.name, 1, 1);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("importing an empty string by text into an empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await importText("");

  await expectCardSlotToNotExist(1);
});

test("the placeholder text of the text importer", async () => {
  server.use(sampleCards, ...defaultHandlers);
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await openImportTextModal();
  await waitFor(() =>
    expect(
      within(screen.getByTestId("import-text")).getByRole("textbox")
    ).toHaveAttribute(
      "placeholder",
      `4x ${cardDocument1.name}
4x ${cardDocument2.name}
4x ${cardDocument3.name}
4x ${cardDocument4.name}

4x t:${cardDocument6.name}

4x b:${cardDocument5.name}`
    )
  );
});

test("the textbox should clear itself after submitting a list", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import a card
  await importText("my search query");

  // a card slot should have been created
  await expectCardSlotToExist(1);

  // open the window again and assert that the textbox is empty
  const textArea = await openImportTextModal();
  expect(textArea.textContent).toEqual("");
});

test("the textbox should not clear itself until the form has been submitted", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // open the window and set some text
  const textArea = await openImportTextModal();
  const user = userEvent.setup();
  await user.clear(textArea);
  await user.type(textArea, "big test");

  // close the window and reopen it
  screen.getByLabelText("import-text-close").click();
  const textArea2 = await openImportTextModal();
  expect(textArea2.textContent).toEqual("big test");
});

test("pressing ctrl+enter while focused on the textarea should submit the form", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import a card
  const textArea = await openImportTextModal();
  const user = userEvent.setup();
  await user.clear(textArea);
  await user.type(textArea, "my search query");
  await user.keyboard("{Control>}{Enter}");

  // a card slot should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});
