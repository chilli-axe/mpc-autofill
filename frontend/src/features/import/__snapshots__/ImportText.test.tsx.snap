// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of text importer 1`] = `
<div
  class="modal-dialog modal-dialog-scrollable"
  data-testid="import-text"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Add Cards — Text
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      <p>
        Type the names of the cards you'd like to add to your order and hit 
        <b>
          Submit
        </b>
        . One card per line.
      </p>
      <div
        class="accordion"
      >
        <div
          class="accordion-item"
        >
          <h2
            class="accordion-header"
          >
            <button
              aria-expanded="false"
              class="accordion-button collapsed"
              type="button"
            >
              Syntax Guide
            </button>
          </h2>
          <div
            class="accordion-collapse collapse"
          >
            <div
              class="accordion-body"
            >
              <ul>
                <li>
                  There are three types of images in 
                  MPC Autofill
                   —
                   
                  card
                  s, 
                  cardback
                  s, and
                   
                  token
                  s. If you search for a card, the search results 
                  <b>
                    won't contain cardbacks or tokens
                  </b>
                  .
                </li>
                <li>
                  <b>
                    Card
                    s
                  </b>
                   are searched for by default.
                </li>
                <li>
                  Search for 
                  <b>
                    token
                    s
                  </b>
                   by putting
                   
                  <code>
                    t:
                  </code>
                   at the start of the query — for example,
                   
                  <code>
                    t:
                    your
                     
                    token
                     name
                  </code>
                  .
                </li>
                <li>
                  Search for 
                  <b>
                    cardback
                    s
                  </b>
                   by putting
                   
                  <code>
                    b:
                  </code>
                   at the start of the query — for example,
                   
                  <code>
                    b:
                    your
                     
                    cardback
                     name
                  </code>
                  .
                </li>
                <li>
                  You may optionally specify the image ID to select by typing your search query, 
                  <code>
                    @
                  </code>
                  , then the image ID — for example,
                   
                  <code>
                    your 
                    card
                     name
                    @
                    1c4M-sK9gd0Xju0NXCPtqeTW_DQTldVU5
                  </code>
                  .
                </li>
                <li>
                  You may specify queries for both the front and the back by separating them with 
                  <code>
                    //
                  </code>
                   — for example,
                   
                  <code>
                    4x goblin 
                    //
                     
                    t:
                    elf
                  </code>
                  .
                </li>
                <li>
                  If you don't specify a back query and your front query is a double-faced card, we will automatically query the back card for you.
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
      <br />
      <form
        class=""
        id="importTextForm"
      >
        <div
          class="mb-3"
        >
          <textarea
            aria-label="import-text"
            class="form-control"
            placeholder=""
            required=""
            rows="12"
          />
        </div>
      </form>
      <p
        class="text-sm-end my-0"
      >
        <i>
          Hint: Submit with Control+Enter.
        </i>
      </p>
    </div>
    <div
      class="modal-footer"
    >
      <button
        aria-label="import-text-close"
        class="btn btn-secondary"
        type="button"
      >
        Close
      </button>
      <button
        aria-label="import-text-submit"
        class="btn btn-primary"
        form="importTextForm"
        type="submit"
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;
